api/types/src/transaction.rs

```diff
@@ -70,9 +70,21 @@ pub enum TransactionData {
     Pending(Box<SignedTransaction>),
 }
 
-impl From<TransactionOnChainData> for TransactionData {
-    fn from(txn: TransactionOnChainData) -> Self {
-        Self::OnChain(txn)
+impl TransactionData {
+    pub fn from_transaction_onchain_data(
+        txn: TransactionOnChainData,
+        latest_ledger_version: u64,
+    ) -> Result<Self> {
+        if txn.version > latest_ledger_version {
+            match txn.transaction {
+                aptos_types::transaction::Transaction::UserTransaction(txn) => {
+                    Ok(Self::Pending(Box::new(txn)))
+                },
+                _ => bail!("Cannot convert non-user transaction to pending"),
```

Comment:
can you double check if this error is handled correctly at all the caller side?

Reply:
this is the place that shouldn't be reached. I don't want to panic so use bail here. updated the message to make it clear.